ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl/Pure.hs

```diff
@@ -0,0 +1,341 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Pure side of the Mempool implementation.
+--
+-- Operations are performed in a pure style returning data types that model
+-- the control flow through the operation and can then be interpreted to perform
+-- the actual STM/IO operations.
+
+module Ouroboros.Consensus.Mempool.Impl.Pure (
+    -- * Mempool
+    pureGetSnapshotFor
+  , pureRemoveTxs
+  , pureSyncWithLedger
+  , pureTryAddTxs
+  , runRemoveTxs
+  , runSyncWithLedger
+  , runTryAddTxs
+    -- * MempoolSnapshot
+  , implSnapshotFromIS
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Maybe (isJust, isNothing)
+import qualified Data.Set as Set
+import           Data.Word (Word32)
+
+import           Control.Tracer
+import           Ouroboros.Consensus.HeaderValidation
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Ledger.SupportsMempool
+import           Ouroboros.Consensus.Mempool.API
+import           Ouroboros.Consensus.Mempool.Impl.Types
+import           Ouroboros.Consensus.Mempool.TxSeq (TicketNo, TxTicket (..),
+                     zeroTicketNo)
+import qualified Ouroboros.Consensus.Mempool.TxSeq as TxSeq
+import           Ouroboros.Consensus.Util (whenJust)
+import           Ouroboros.Consensus.Util.IOLike
+
+{-------------------------------------------------------------------------------
+  Mempool Implementation
+-------------------------------------------------------------------------------}
+
+-- | Data constructor containing the control flow for 'tryAddTxs'. Can be
+-- interpreted in an IOLike monad with 'runTryAddTxs'.
+data TryAddTxs blk =
+    -- | No more transactions to add or the mempool reached its full capacity.
+    Done [GenTx blk]
+    -- | There are more transactions to process.
+  | More
+      (Maybe (InternalState blk))
+      -- ^ If the transaction was accepted, the new state that can be written to
+      -- the TVar.
+      (TraceEventMempool blk)
+      -- ^ The event emitted by the operation.
+      (MempoolAddTxResult blk)
+      -- ^ The result of trying to add the transaction to the mempool.
+      (InternalState blk -> TryAddTxs blk)
```

Comment:
This last argument is a non-trivial maneuver. I'm wondering if we can simplify by being a bit more direct.

```
...
tryAddTxs      = implTryAddTxs istate
...

implTryAddTxs istate txs =
    go []
  where
    go acc = \case
      []     -> pure (reverse acc, [])
      tx:txs -> join $ atomically $ do
        is <- readTVar istate
        -- this case is essentially psuedo-code; missing eg tracer logic
        pureTryAddTx {- NOTE one tx at a time -} ... is tx >>= \case
          Nothing            -> pure $ pure (reverse acc, tx:txs)
          Just (is', result) -> do writeTVar istate is'; pure $ go (result:acc) txs
```

The pure code handles a single tx at a time, while the monadic wrapper handles one tx at a time, only maintaining the list of results and the list of unattempted txs.

This way `TryAddTxs` doesn't even have to be recursive, let alone a coroutine sort of thing.